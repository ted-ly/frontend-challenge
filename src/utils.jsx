import { useState, useCallback } from 'react';

export const initialState = {
    name: '',
    password: '',
    color: '',
    email: '',
    terms: false,
};

export const onClickNext = (to, setPathname) => {
    setPathname(to);
    window.history.pushState(null, '', to);
};

export const onClickBack = (from, setPathname) => {
    setPathname(from);
    window.history.pushState(null, '', from);
}; 

export const onSubmit = async (formData, setPathname) => {
    const resp = await fetch('http://localhost:3001/api/submit', { method: 'POST', headers: {'Content-Type': 'application/json'}, body: JSON.stringify(formData) });
    if (resp.status >= 400) {
        window.history.pushState(null, '', '/error');
        setPathname('/error');
    } else {
        window.history.pushState(null, '', '/success');
        setPathname('/success');
    }
};

export const createOnClickRestart = ({ setFormData, setPathname }) => useCallback(() => {
    window.history.pushState(null, '', '/');
    setPathname('/');
    setFormData({...initialState});
}, []);

export const createOnFormDataChange = ({ setFormData, formData }) => useCallback((e) => {
    const name = e.target.name;
    const value = e.target.value;
    const newFormData = {
        ...formData,
        [name]: name ==='terms' ? !formData.terms : value,
    };

    setFormData({...newFormData});
}, [formData]); 

export const useForm = () => {
    const [formData, setFormData] = useState(initialState);
    const [pathname, setPathname] = useState('/');
    const onFormDataChange = createOnFormDataChange({ formData, setFormData });
    const onClickRestart = createOnClickRestart({ setFormData, setPathname });

    return { onClickRestart, onFormDataChange, formData, setPathname };
};

export const toCapitalized = (str) => {
    return str.slice(0, 1).toUpperCase() + str.slice(1);
}

export const toTransformText = (text, key) => {
    if (key === 'terms') {
        return text ? 'Agreed' : 'Disagreed';
    }

    if (key === 'password') {
        return '*'.repeat(text.length);
    }

    return text;
}; 