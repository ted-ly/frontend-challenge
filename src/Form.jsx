import React from 'react';
import AdditionalInfo from './AdditionalInfo';
import Confirmation from './Confirmation';
import PostSubmit from './PostSubmit';
import Route from './Route';
import Signup from './Signup';
import { useForm } from './utils';

export default function Form() {
    const dataProps = useForm();

    return (
        <div className='form'>
            <Route path='/'><Signup {...dataProps} /></Route>
            <Route path='/more-info'><AdditionalInfo {...dataProps} /></Route>
            <Route path='/confirmation'><Confirmation {...dataProps} onClickNext={dataProps.onSubmit} /></Route>
            <Route path='/error'><PostSubmit {...dataProps} isError={true} /></Route>
            <Route path='/success'><PostSubmit {...dataProps} /></Route>
        </div>
    );
}