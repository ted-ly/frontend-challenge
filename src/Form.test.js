import renderer from 'react-test-renderer';
import Form from './Form';
import { initialState } from './utils';

describe('Form', () => {
    it('should match snapshot', () => {
        const component = renderer.create(<Form formData={initialState} />);
        expect(component.toJSON()).toMatchSnapshot();
    });
});