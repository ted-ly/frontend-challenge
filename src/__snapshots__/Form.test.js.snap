// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form should match snapshot 1`] = `
<div
  className="form"
>
  <div
    className="step-container"
  >
    <div
      className="step-title"
    >
      Sign Up
    </div>
    <div
      className="signup-container"
    >
      <input
        name="name"
        onChange={[Function]}
        placeholder="First Name"
        type="text"
        value=""
      />
      <input
        name="password"
        onChange={[Function]}
        placeholder="Password"
        type="password"
        value=""
      />
      <input
        name="email"
        onChange={[Function]}
        placeholder="E-mail"
        type="email"
        value=""
      />
      <div
        className="button-container"
      >
        <button
          onClick={[Function]}
          type="button"
        >
          Next
        </button>
      </div>
    </div>
  </div>
</div>
`;
